include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -0,0 +1,835 @@
+                    {
+                        __sub_group_partials[__i * __sub_group_size + __sub_group_local_id] =
+                            __tmp_ptr[__subgroups_before_my_group + __i * __sub_group_size + __sub_group_local_id];
+                    }
+
+                    // step 2) load 32, 64, 96, etc. work-group carry outs on every work-group; then
+                    //         compute the prefix in a sub-group to get global work-group carries
+                    //         memory accesses: gather(63, 127, 191, 255, ...)
+                    std::uint32_t __offset = __num_sub_groups_local - 1;
+                    // only need 32 carries for WGs0..WG32, 64 for WGs32..WGs64, etc.
+                    if (__g > 0)
+                    {
+                        // only need the last element from each scan of num_sub_groups_local subgroup reductions
+                        const auto __elements_to_process = __subgroups_before_my_group / __num_sub_groups_local;
+                        const auto __pre_carry_iters =
+                            oneapi::dpl::__internal::__dpl_ceiling_div(__elements_to_process, __sub_group_size);
+                        if (__pre_carry_iters == 1)
+                        {
+                            // single partial scan
+                            auto __proposed_idx = __num_sub_groups_local * __sub_group_local_id + __offset;
+                            auto __remaining_elements = __elements_to_process;
+                            auto __reduction_idx = (__proposed_idx < __subgroups_before_my_group)
+                                                       ? __proposed_idx
+                                                       : __subgroups_before_my_group - 1;
+                            auto __value = __tmp_ptr[__reduction_idx];
+                            __sub_group_scan_partial<__sub_group_size, /*__is_inclusive=*/true,
+                                                     /*__init_present=*/false>(__sub_group, __value, __reduce_op,
+                                                                               __carry_last, __remaining_elements);
+                        }
+                        else
+                        {
+                            // multiple iterations
+                            // first 1 full
+                            auto __value = __tmp_ptr[__num_sub_groups_local * __sub_group_local_id + __offset];
+                            __sub_group_scan<__sub_group_size, /*__is_inclusive=*/true, /*__init_present=*/false>(
+                                __sub_group, __value, __reduce_op, __carry_last);
+
+                            // then some number of full iterations
+                            for (std::uint32_t __i = 1; __i < __pre_carry_iters - 1; __i++)
+                            {
+                                auto __reduction_idx = __i * __num_sub_groups_local * __sub_group_size +
```

Comment:
We might want to create helpers for `__num_sub_groups_local * __sub_group_size` and `__num_sub_groups_local * __sub_group_local_id + __offset` and reuse them in this loop and the partial iteration below.